cassandane/Cassandane/Unit/TestCase.pm

```diff
@@ -329,4 +329,120 @@ sub assert_num_lte
                   "$actual is not less-than-or-equal-to $expected");
 }
 
+{
+  package Cassandane::Test::URL;
+
+  use Plack::Loader;
+  use Plack::Request;
+  use Plack::Response;
+
+  use Test::TCP;
+
+  sub new {
+    my ($pkg, $args) = @_;
+
+    my %attrs;
+
+    for my $required (qw(app)) {
+      $attrs{$required} = delete $args->{$required};
+      unless ($attrs{$required}) {
+        require Carp;
+
+        Carp::croak("'$required' required for Cassandane::Test::URL->new");
+      }
+    }
+
+    my $self = bless {}, __PACKAGE__;
+
+    $self->update($attrs{app});
+
+    return $self;
+  }
+
+  sub url {
+    my ($self) = @_;
+
+    if ($self->was_unregistered) {
+      require Carp;
+      Carp::croak("Cannot call ->url after ->unregister has been called!");
+    }
+
+    $self->{url};
+  }
+
+  sub _guard { shift->{_guard} }
+
+  sub unregister {
+    my ($self) = @_;
+
+    delete $self->{_guard};
+    $self->{was_unregistered} = 1;
+  }
+
+  sub was_unregistered { shift->{was_unregistered} }
+
+  sub update {
+    my ($self, $content_or_app) = @_;
+
+    unless (ref $content_or_app) {
+      my $content = $content_or_app;
+      # Plain ol' successful response
+      $content_or_app = sub {
+        return [
+          200,
+          [],
+          [ $content ],
+        ];
+      };
+    }
+
+    my $app = sub {
+      my $sock_or_port = shift;
+      my $server = Plack::Loader->auto(
+        host => '0.0.0.0',
+        port => $sock_or_port,
+      );
+      $server->run($content_or_app);
+      exit;
```

Comment:
> I don't know Plack or Test::TCP. Are they well-behaved with respect to... actually I don't have a word for this...
> 
> Okay let's imagine this scenario:
> 
>     * I have written a Cassandane test
> 
>     * When the test runs, the entirety of Cassandane, including all its own globals, is in scope
> 
>     * My test uses this to spawn off a fake http server for the Cyrus-under-test to interact with
> 
>     * Eventually, that server calls `exit` here
> 
>     * Is the server's exit going to try to `DESTROY` everything it has in scope, including all of Cassandane's globals?
> 
> 
> I can't see where the fork happens (I'm not about to go reading the Plack or Test::TCP code), so I can't see whether the fork execs a new image (good), or keeps running a clone of the original image (bad). I ask because I've already cleaned up a few messes like this in Cassandane's internals, and there's at least one remaining that I haven't gotten to yet, so I'm wary of another one sliding in unobserved.
> 
> I would like to assume respected modules like Plack/Test::TCP behave well, but I don't _know_ that, so I'm asking. :)

I don't think its on any module that forks to behave well, because there's not an easy to *to* behave well.

Any time perl forks and does not exec, you'll have this problem. I think its on the destructors to ensure they aren't being in run in the process they need to be run in if it can cause problems.

Usually I do this by taking the parent PID when the destructor is set up and refusing to run if the current pid does not match:

```
package Foo;
my $PARENT = $$;

sub DESTROY {
  return if $$ != $PARENT;
}
```

etc etc

You could attempt to have all of your children that fork and don't exec exit using POSIX::_exit(), but you'd have to account for any signal handlers that cause perl to exit normally or die. (IE: If there's an exception, perl will throw a $SIG{__DIE__}, then your _exit() won't be called unless you set up something to catch __DIE__, because even die() calls destructors on exit)

I see the few places where we fork we do something with $SIG{TERM} to call _exit(0). We could add that, because Test::TCP uses $SIG{TERM} when stop() is called or when its guard goes out of scope.

Reply:
We (newly) do something like this already in the one place it's burned us so far: https://github.com/cyrusimap/cyrus-imapd/blob/5fa7876cff6671794412f65e8241b398cbf0e4d8/cassandane/Cassandane/Instance.pm#L1538-L1541

But Cassandane is big and thorny and I'm sure there's many more delicate bits, they just haven't been kicked yet.

The two(?) things I'm aware of that still fork-without-exec are `Cassandane::Instance::_start_smtpd` and `Cassandane::Instance::start_httpd`, and I have littered both with comments reminding myself to rip them out into standalone scripts that can be fork-and-exec'd (or started by Cyrus rather than by Cassandane), like I already did with `utils/fake*d`.

It's interesting to dig around, because now that I have, I think `start_httpd` could be replaced by this thing you're making.  For that matter, in hindsight I think `start_httpd` is what I was thinking of when I told Ken that I thought we already had something like this.

> I see the few places where we fork we do something with `$SIG{TERM}` to call `_exit(0)`. We could add that, [...]

If those "few places" are `_start_smtp` and `start_httpd`, then I consider them terrible models we shouldn't copy ;)

In the case of this thing (and `start_httpd`), where the behaviour of the spawned server is defined by a sub provided by the caller, I'm not sure off the top of my head how to handle it cleanly.  It's easier for the other two, cause one's behaviour is static, and the other's is defined by a file.

The alternative -- protecting DESTROY -- is awkward since Cassandane expects that it's controlling all the forking (except for that done by Cyrus), and we'd probably need to do this in every extant file if we're relinquishing that.

Reply:
Okay so I've added back in a pattern you were dubious about before, but I think in the end its worth it here (that is, declaring a tiny package within the same file). See https://github.com/cyrusimap/cyrus-imapd/pull/3857/commits/d789a2fc3ed5be300743f3a46d0358e49c30340a

What it does is create an object that calls _exit() when it goes out of scope, which is, as far as I can tell, always guaranteed to happen during the run phase of perl, which means it will happen when this block ends, before any other END/DESTROY block gets to run, effectively preventing any other cleanup code from ever running.

What do you think?

Reply:
Oh that's cute.  My pure perl isn't sharp enough to anticipate any edge cases you probably haven't yourself, but this seems like it should play nice with Cassandane.

I think the embedded package thing bothers me a lot less when the whole thing is small enough to fit on screen at once (even in the worst case of a 24-row terminal).  By contrast, Cassandane/Unit/TestPlan.pm has a bunch of multi-hundred-line packages in the same file and it's a nightmare to keep track of what's going on.

I'm gonna pull this branch and build/run it to see what happens, though I don't really expect we'd see weird edge cases turn up until we start using it heavily.  I think the main thing I'll be looking for (assuming it passes) is that I don't see any "ignoring DESTROY from invalid caller" (or whatever it was) in the logs.

Reply:
Of course I won't see any "ignoring DESTROY...", because the test isn't one that starts up any Instances at all, hah.  Hmm